src/App.tsx

```diff
@@ -14,11 +18,125 @@ const goodsFromServer: string[] = [
   'Garlic',
 ];
 
-const App: React.FC = () => (
-  <div className="App">
-    <h1>Goods</h1>
-    {goodsFromServer.length}
-  </div>
-);
+type State = {
+  goods: string[];
+  buttonVisible: boolean;
+  isReversed: boolean;
+  sortBy: string;
+};
+
+type Props = {};
+
+enum Sortby {
+  Name = 'name',
+  Length = 'length',
+}
+
+class App extends React.Component<Props, State> {
+  state: State = {
+    goods: [...goodsFromServer],
+    buttonVisible: true,
+    isReversed: false,
+    sortBy: '',
+  };
+
+  startHandler = () => {
+    this.setState((state) => ({
+      buttonVisible: !state.buttonVisible,
+    }));
+  };
+
+  reverseHandler = () => {
+    this.setState((state) => ({
+      isReversed: !state.isReversed,
+    }));
+  };
+
+  sortByNameHandler = () => {
+    this.setState(() => ({
+      sortBy: Sortby.Name,
+      isReversed: false,
+    }));
+  };
+
+  sortByLengthHandler = () => {
+    this.setState(() => ({
+      sortBy: Sortby.Length,
+      isReversed: false,
+    }));
+  };
+
+  resetHandler = () => {
+    this.setState(() => ({
+      isReversed: false,
+      sortBy: '',
+    }));
+  };
+
+  render() {
+    const {
+      buttonVisible, goods, isReversed, sortBy,
+    } = this.state;
+    const isButtonVisible = { buttonVisible: buttonVisible === true };
+    const initialGoods = [...goods];
+
+    if (sortBy === Sortby.Name) {
+      initialGoods.sort((a, b) => a.localeCompare(b));
+    }
+
+    if (sortBy === Sortby.Length) {
+      initialGoods.sort((a, b) => a.length - b.length);
+    }
+
+    if (isReversed) {
+      initialGoods.reverse();
+    }
```

Comment:
Create a separate function `getVisibleGoods` for this logic

Reply:
for this { buttonVisible: buttonVisible === true } ?